src/io/flutter/dart/DartAnalysisServerServiceEx.java

```diff
@@ -0,0 +1,145 @@
+public class DartAnalysisServerServiceEx {
+  public final DartAnalysisServerService base;
+
+  private final AnalysisServer analysisServer;
+  private final RequestSink requestSink;
+  private final List<DartAnalysisServerServiceExResponseListener> listeners = new ArrayList<>();
+
+  public DartAnalysisServerServiceEx(DartAnalysisServerService base, AnalysisServer analysisServer) {
+    this.base = base;
+    this.analysisServer = analysisServer;
+    this.requestSink = ReflectionUtil.getField(analysisServer.getClass(), analysisServer, RequestSink.class, "requestSink");
+  }
+
+  public void addListener(DartAnalysisServerServiceExResponseListener listener) {
+    synchronized (listeners) {
+      if (!listeners.contains(listener)) {
+        listeners.add(listener);
+      }
+    }
+  }
+
+  public void removeListener(DartAnalysisServerServiceExResponseListener listener) {
+    synchronized (listeners) {
+      listeners.remove(listener);
+    }
+  }
+
+  @Nullable
+  public String generateUniqueId() {
+    try {
+      final Method generateUniqueId = ReflectionUtil.getMethod(analysisServer.getClass(), "generateUniqueId");
+      if (generateUniqueId != null) {
+        return (String)generateUniqueId.invoke(analysisServer);
+      }
+    }
+    catch (IllegalAccessException | InvocationTargetException ignored) {
+    }
+    return null;
+  }
+
+  public void sendRequest(JsonObject request) {
+    requestSink.add(request);
+  }
+
+  private class ResponseStreamEx implements ResponseStream {
+    final ResponseStream base;
+
+    ResponseStreamEx(ResponseStream base) {
+      this.base = base;
+    }
+
+    @Override
+    public void lastRequestProcessed() {
+      base.lastRequestProcessed();
+    }
+
+    @Override
+    public JsonObject take() throws Exception {
+      final JsonObject json = base.take();
+      synchronized (listeners) {
+        for (DartAnalysisServerServiceExResponseListener listener : ImmutableList.copyOf(listeners)) {
+          listener.onResponse(json);
+        }
+      }
+      return json;
+    }
+  }
+
+  private static final WeakHashMap<AnalysisServer, DartAnalysisServerServiceEx> serverToServiceEx = new WeakHashMap<>();
+
+  public static DartAnalysisServerServiceEx get(DartAnalysisServerService base) {
+    if (base == null) {
+      return null;
+    }
+
+    final AnalysisServer analysisServer =
+      ReflectionUtil.getField(DartAnalysisServerService.class, base, AnalysisServer.class, "myServer");
+    if (analysisServer == null) {
+      return null;
+    }
+
+    {
+      final DartAnalysisServerServiceEx existingEx = serverToServiceEx.get(analysisServer);
+      if (existingEx != null) {
+        return existingEx;
+      }
+    }
+
+    final ResponseStream serverResponseStream =
+      ReflectionUtil.getField(analysisServer.getClass(), analysisServer, ResponseStream.class, "responseStream");
+    if (serverResponseStream == null) {
+      return null;
+    }
+
+    Thread responseReaderThread = null;
```

Comment:
Can you describe what this block of code is doing in a comment? I think it'll be useful for other people through the code.